src/services/registrationData.ts

```diff
@@ -103,22 +103,173 @@ export async function upsertRegistrationData({
+ * @param filteredRegistrationData - Filtered registration data.
+ * @returns Combined registration data.
+ */
+function mapToCombinedData(
+  filteredRegistrationData: {
+    id: string;
+    registrationFieldId: string;
+    registrationId: string;
+    value: string;
+  }[],
+  registrationFields: {
+    registrationFieldId: string;
+    questionId: string;
+    questionOptionType: string;
+  }[],
+): {
+  registrationId: string;
+  questionId: string;
+  values: { [questionOptionType: string]: string };
+}[] {
+  const combinedData = filteredRegistrationData.map((data) => {
+    const matchingField = registrationFields.find(
+      (field) => field.registrationFieldId === data.registrationFieldId,
+    );
+    if (matchingField) {
+      return {
+        registrationId: data.registrationId,
+        questionId: matchingField.questionId,
+        values: {
+          [matchingField.questionOptionType]: data.value,
+        },
+      };
+    }
+    throw new Error(`No matching field found for registrationFieldId: ${data.registrationFieldId}`);
+  });
+
+  return combinedData;
+}
+
+/**
+ * Updates or inserts question options based on the combined data.
+ * @param dbPool - The database pool instance.
+ * @param combinedData - Combined registration data.
+ */
+async function updateQuestionOptions(
+  dbPool: PostgresJsDatabase<typeof db>,
+  combinedData: {
+    registrationId: string;
+    questionId: string;
+    values: { [questionOptionType: string]: string };
+  }[],
+): Promise<void> {
+  for (const data of combinedData) {
+    // Check whether a corresponding question option exists
+    const existingQuestionOption = await dbPool.query.questionOptions.findFirst({
+      where: eq(db.questionOptions.registrationId, data.registrationId),
+    });
+
+    if (existingQuestionOption) {
+      // Update question option
+      await dbPool
+        .update(db.questionOptions)
+        .set({
+          registrationId: data.registrationId,
+          questionId: data.questionId,
+          optionTitle: data.values['TITLE'] || existingQuestionOption.optionTitle,
+          optionSubTitle: data.values['SUBTITLE'] || existingQuestionOption.optionSubTitle,
+          updatedAt: new Date(),
+        })
+        .where(eq(db.questionOptions.id, existingQuestionOption.id))
+        .returning();
+    } else {
+      // Insert new question option
+      await dbPool
+        .insert(db.questionOptions)
+        .values({
+          registrationId: data.registrationId,
+          questionId: data.questionId,
+          optionTitle: data.values['TITLE'] || '',
+          optionSubTitle: data.values['SUBTITLE'] || '',
+          createdAt: new Date(),
+          updatedAt: new Date(),
+        })
+        .returning();
+    }
+  }
+}
+
+/**
+ * Main function to update question options.
+ * @param dbPool - The database pool instance.
+ * @param registrationData - An array of registration data.
+ */
+export async function updateQuestionOptionsMain(
```

Comment:
rename to upsertQuestionOptionFromRegistrationData

Reply:
done